modules/schema_validation.lua

```diff
@@ -35,315 +26,264 @@ local function dump(o)
+       return true
+      else 
+         return false, "Parameter ".. nameofParameter..": got "..elem1..", expected "..schemaElem
       end
-    elseif xmlNode.array == 'true' then
-      if (type(dataNode) == 'table' ) then
-        for _,arrElement in ipairs(dataNode) do
-          while true do
-            if (xmlNode.class == 'Enum' and types.enum[xmlNode['type'] ][arrElement] ) then break
-            elseif (xmlNode.class == 'Struct' and types.struct[xmlNode['type'] ][arrElement] ) then break
-            elseif(compare_type(string.lower(type(arrElement)), xmlNode.class)) then break
-            else
-              bool_result = false
-              errorMessage[ key ] = "not valid type: into "..key .. " " .. string.lower(type(arrElement)) .. " " .. tostring(xmlNode.class)
-              break
-            end
-          end
-        end
-      else
-        bool_result = false
-        errorMessage[ key ] = "expected: '" .. key .."' with type [ array ]"
+    end
+    local interface_name, complex_elem_name = GetNames(schemaElem)
+    if self.schema.interface[interface_name].enum[complex_elem_name] ~= nil then
+      if(self.schema.interface[interface_name].enum[complex_elem_name][data_elem] ~= nil) then 
+        return true
+        else 
+          return false, "Parameter ".. nameofParameter.. ": got "..elem1..", expected enum value"
       end
-    else
-      if (types.classes[xmlNode.class] ~= 'nil') then
-        if (not compare_type(string.lower(type(dataNode)), xmlNode.class) ) then
-          bool_result = false
-          errorMessage[ key ] = "not valid type '"..key .."': current type - '".. string.lower(type(dataNode)) .."' ; Available - '".. tostring(xmlNode.class) .."'"
-        end
+    end
+    if self.schema.interface[interface_name].struct[complex_elem_name] ~= nil then
+      if (elem1~='table') then 
+        return false, "Parameter ".. nameofParameter.. ": got "..elem1..", expected struct"
       else
-        errorMessage[ key ] = "not valid type: "..key
+        return self:CompareStructs(data_elem, self.schema.interface[interface_name].struct[complex_elem_name])
       end
     end
   end
+  return false, "Parameter ".. nameofParameter..": got "..elem1..", expected "..schemaElem
+end
+
+
+local function CountInArray(T)
+  local count = 0
+  for _ in pairs(T) do count = count + 1 end
+  return count
+end
 
-  local function schemaCompare(xml_table,user_data)
-    if (type(xml_table)~="table") then
-      return nil, "Empty Data"
+local function CheckArray(data_elem, schemaElem)
+  if json.isArray(data_elem) then
+    arraySize = CountInArray(data_elem)
+    if(arraySize < schemaElem["minsize"]) then
+      return false, "in array get size: "..arraySize..", expected minsize:"..schemaElem["minsize"]
+    end
+    if(arraySize > schemaElem["maxsize"]) then
+      return false, "in array get size: "..arraySize..", expected maxsize:"..schemaElem["maxsize"]
     end
+  else 
+    return false
+  end
+  return true
```

Comment:
@VVeremjova  Please try to rewrite this method fore better readability. AL Lease move this return inside if.
Or invert if and avoid else.